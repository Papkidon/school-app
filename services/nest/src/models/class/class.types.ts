import { z } from 'zod';

export const classCreateSchema = z
  .object({
    data: z.object({
      number: z.number().int().min(0),
      max_population: z.number().int().min(0),
    }),
  })
  .strict();

export const classUpdateSchema = classCreateSchema.deepPartial();

export type classesWithSubjectsDto = {
  id: string;
  number: number;
  max_population: number;
  ClassSubject?: [
    {
      class_id: string;
      subject_id: string;
      subject?: {
        id: string;
        name: string;
        points: number;
      };
    },
  ];
}[];

export type mappedClassesWithSubjects = {
  id: string;
  number: number;
  max_population: number;
  subjects: {
    id: string;
    name: string;
    points: number;
  }[];
}[];

export type classType = {
  data: {
    number: number;
    max_population: number;
  };
};
