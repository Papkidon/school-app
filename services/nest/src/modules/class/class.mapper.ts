import {
  classesWithSubjectsDto,
  mappedClassesWithSubjects,
} from '../../models/class/class.types';

export const mapClassesWithSubjects = (
  classes: classesWithSubjectsDto,
): mappedClassesWithSubjects => {
  if (!classes.length) return [];
  return classes.map((subject) => {
    return {
      id: subject.id,
      number: subject.number,
      max_population: subject.max_population,
      subjects: subject.ClassSubject.map((clas) => {
        return {
          id: clas.subject.id,
          name: clas.subject.name,
          points: clas.subject.points,
        };
      }),
    };
  });
};
