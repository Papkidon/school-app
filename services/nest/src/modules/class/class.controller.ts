import {
  Body,
  Controller,
  Delete,
  Get,
  HttpException,
  HttpStatus,
  Param,
  Patch,
  Post,
} from '@nestjs/common';
import { Class } from '@prisma/client';
import {
  classCreateSchema,
  classType,
  classUpdateSchema,
} from '../../models/class/class.types';
import { DeepPartial, idSchema } from '../../models/utility.types';
import { ZodValidationPipe } from '../../pipes/zod.validation.pipe';
import { mapClassesWithSubjects } from './class.mapper';
import { ClassService } from './class.service';

@Controller('api/v1/class')
export class ClassController {
  constructor(private readonly classService: ClassService) {}

  @Get('')
  async findAll(): Promise<Class[]> {
    try {
      const result = await this.classService.findAll();

      if (!Object.keys(result).length) {
        throw new HttpException(
          'No class data exists in the database',
          HttpStatus.NOT_FOUND,
        );
      }

      return result;
    } catch (error) {
      throw error;
    }
  }

  @Get('subjects')
  async findAllClassesWithSubjects() {
    try {
      const result = mapClassesWithSubjects(
        await this.classService.findAllClassesWithSubjects(),
      );

      if (!result.length) {
        throw new HttpException(
          'No class data exists in the database',
          HttpStatus.NOT_FOUND,
        );
      }

      return result;
    } catch (error) {
      throw error;
    }
  }

  @Post('')
  async create(
    @Body(new ZodValidationPipe(classCreateSchema)) createClassDto: classType,
  ): Promise<Class> {
    try {
      const foundClass = await this.classService.findByNumber(
        createClassDto.data.number,
      );

      if (foundClass) {
        throw new HttpException(
          'Class with given number already exists',
          HttpStatus.CONFLICT,
        );
      }

      return await this.classService.create(createClassDto);
    } catch (error) {
      throw error;
    }
  }

  @Get(':id')
  async findById(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
  ): Promise<Class> {
    try {
      const result = await this.classService.findById(id);

      if (!result) {
        throw new HttpException(
          'User with given id does not exist in the database',
          HttpStatus.NOT_FOUND,
        );
      }

      return result;
    } catch (error) {
      throw error;
    }
  }

  @Patch(':id')
  async update(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
    @Body(new ZodValidationPipe(classUpdateSchema))
    body: DeepPartial<classType>,
  ): Promise<Class> {
    try {
      const foundClass = await this.classService.findById(id);

      if (!foundClass) {
        throw new HttpException(
          'Class with given id does not exist in the database',
          HttpStatus.NOT_FOUND,
        );
      }

      return await this.classService.update(id, body);
    } catch (error) {
      throw error;
    }
  }

  @Delete(':id')
  async delete(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
  ): Promise<Class> {
    try {
      const foundClass = await this.classService.findById(id);

      if (!foundClass) {
        throw new HttpException(
          'Class with given id does not exist in the database',
          HttpStatus.NOT_FOUND,
        );
      }

      return await this.classService.delete(id);
    } catch (error) {
      throw error;
    }
  }
}
