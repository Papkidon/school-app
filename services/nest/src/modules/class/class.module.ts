import { Module } from '@nestjs/common';
import { ClassController } from './class.controller';
import { ClassService } from './class.service';
import { ClassRepository } from './class.repository';
import { LoggerModule } from 'nestjs-pino';
import { PrismaModule, PrismaService } from 'nestjs-prisma';

@Module({
  imports: [
    LoggerModule.forRoot({
      pinoHttp: {
        transport: {
          target: 'pino-pretty',
          options: {
            singleLine: true,
          },
        },
      },
    }),
    PrismaModule.forRoot(),
  ],
  controllers: [ClassController],
  providers: [ClassService, ClassRepository, PrismaService],
})
export class ClassModule {}
