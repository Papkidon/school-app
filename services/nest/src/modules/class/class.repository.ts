import { Injectable } from '@nestjs/common';
import { PrismaService as PrismaService } from 'nestjs-prisma';
import { classType } from '../../models/class/class.types';
import { DeepPartial } from '../../models/utility.types';
import { Class } from '@prisma/client';
import { classesWithSubjectsDto } from '../../models/class/class.types';

@Injectable()
export class ClassRepository {
  constructor(private prisma: PrismaService) {}
  async findAll(): Promise<Class[]> {
    return await this.prisma.class.findMany();
  }

  async findById(id: string): Promise<Class> {
    return await this.prisma.class.findUnique({ where: { id: id } });
  }

  async findAllClassesWithSubjects(): Promise<classesWithSubjectsDto> {
    return (await this.prisma.class.findMany({
      include: {
        ClassSubject: {
          include: {
            subject: true,
          },
        },
      },
    })) as unknown as classesWithSubjectsDto;
  }

  async create(body: any): Promise<Class> {
    return await this.prisma.class.create(body);
  }

  async delete(id: string): Promise<Class> {
    return await this.prisma.class.delete({ where: { id: id } });
  }

  async update(id: string, body: DeepPartial<classType>): Promise<Class> {
    return await this.prisma.class.update({
      where: { id: id },
      data: { ...body.data },
    });
  }
}
