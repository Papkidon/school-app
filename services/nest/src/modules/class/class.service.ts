import { Injectable } from '@nestjs/common';
import { ClassRepository } from './class.repository';
import { classType } from '../../models/class/class.types';
import { DeepPartial } from '../../models/utility.types';
import { Class } from '@prisma/client';
import { classesWithSubjectsDto } from '../../models/class/class.types';

@Injectable()
export class ClassService {
  constructor(private classRepository: ClassRepository) {}
  async findAll(): Promise<Class[]> {
    return await this.classRepository.findAll();
  }

  async findById(id: string): Promise<Class> {
    return await this.classRepository.findById(id);
  }

  async findByNumber(number: number): Promise<Class> {
    return await this.classRepository.findByNumber(number);
  }

  async findAllClassesWithSubjects(): Promise<classesWithSubjectsDto> {
    return await this.classRepository.findAllClassesWithSubjects();
  }

  async create(body: any): Promise<Class> {
    return await this.classRepository.create(body);
  }

  async delete(id: string): Promise<Class> {
    return await this.classRepository.delete(id);
  }

  async update(id: string, body: DeepPartial<classType>): Promise<Class> {
    return await this.classRepository.update(id, body);
  }
}
