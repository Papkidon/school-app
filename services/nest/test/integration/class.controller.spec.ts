import { INestApplication } from '@nestjs/common';
import { Test } from '@nestjs/testing';
import * as request from 'supertest';
import { classesWithSubjectsDto } from '../../src/models/class/class.types';
import { ClassModule } from '../../src/modules/class/class.module';
import { ClassRepository } from '../../src/modules/class/class.repository';
import {
  dummyClassCreate,
  dummyClassesWithSubjects,
} from './dummies/dummy.data';

describe('[ClassController] Integration tests', () => {
  let app: INestApplication;
  const classRepository = {
    findAll: () => ['dummyFoundClasses'],
    findById: () => 'dummyFoundClassById',
    findByNumber: () => 'dummyFoundClassByNumber',
    findAllClassesWithSubjects: () =>
      dummyClassesWithSubjects as unknown as classesWithSubjectsDto,
    create: () => 'dummyCreatedClass',
    update: () => 'dummyUpdatedClass',
    delete: () => 'dummyDeletedClass',
  };

  beforeAll(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [ClassModule],
    })
      .overrideProvider(ClassRepository)
      .useValue(classRepository)
      .compile();

    app = moduleRef.createNestApplication();
    await app.init();
  });

  it(`Get all classes - should pass when at least one class exists in the db`, async () => {
    const response = await request(app.getHttpServer()).get('/api/v1/class/');

    expect(response.status).toEqual(200);
    expect(response.body).toMatchSnapshot();
  });

  it(`Get all classes - should fail when no class data exists in the db`, async () => {
    jest.spyOn(classRepository, 'findAll').mockImplementationOnce(() => []);

    const response = await request(app.getHttpServer()).get('/api/v1/class/');

    expect(response.status).toEqual(404);
    expect(response.body).toMatchSnapshot();
  });

  it(`Get class by id - should pass when class with given id exists in the db`, async () => {
    const response = await request(app.getHttpServer()).get(
      '/api/v1/class/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    );

    expect(response.status).toEqual(200);
    expect(response.body).toMatchSnapshot();
  });

  it(`Get class by id - should fail when class with given id does not exists in the db`, async () => {
    jest.spyOn(classRepository, 'findById').mockImplementationOnce(() => '');

    const response = await request(app.getHttpServer()).get(
      '/api/v1/class/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    );

    expect(response.status).toEqual(404);
    expect(response.body).toMatchSnapshot();
  });

  it('Get class by id - should fail when id parameter is not UUID', async () => {
    const response = await request(app.getHttpServer()).get(
      '/api/v1/class/123',
    );

    expect(response.status).toEqual(422);
    expect(response.body).toMatchSnapshot();
  });

  it('Get classes with subjects - should pass when data exists', async () => {
    const response = await request(app.getHttpServer()).get(
      '/api/v1/class/subjects',
    );

    expect(response.status).toEqual(200);
    expect(response.body).toMatchSnapshot();
  });

  it('Get classes with subjects - should fail when data does not exists', async () => {
    jest
      .spyOn(classRepository, 'findAllClassesWithSubjects')
      .mockImplementationOnce(() => '' as unknown as classesWithSubjectsDto);

    const response = await request(app.getHttpServer()).get(
      '/api/v1/class/subjects',
    );

    expect(response.status).toEqual(404);
    expect(response.body).toMatchSnapshot();
  });

  it(`Create class - should pass when request body is correct`, async () => {
    jest
      .spyOn(classRepository, 'findByNumber')
      .mockImplementationOnce(() => '');

    const response = await request(app.getHttpServer())
      .post('/api/v1/class/')
      .send(dummyClassCreate);

    expect(response.status).toEqual(201);
    expect(response.body).toMatchSnapshot();
  });

  it(`Create class - should fail when class with given number already exists`, async () => {
    const response = await request(app.getHttpServer())
      .post('/api/v1/class/')
      .send(dummyClassCreate);

    expect(response.status).toEqual(409);
    expect(response.body).toMatchSnapshot();
  });

  it('Create class - should fail when request body is incorrect', async () => {
    const response = await request(app.getHttpServer())
      .post('/api/v1/class/')
      .send({ wrong: 'body' });

    expect(response.status).toEqual(422);
    expect(response.body).toMatchSnapshot();
  });

  it(`Update class - should pass when class with given id exists in the db`, async () => {
    jest
      .spyOn(classRepository, 'findById')
      .mockImplementationOnce(() => 'dummyFoundClassToUpdate');

    const response = await request(app.getHttpServer()).patch(
      '/api/v1/class/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    ); // It passes without the body, because every property is optional

    expect(response.status).toEqual(200);
    expect(response.body).toMatchSnapshot();
  });

  it(`Update class - should fail when class with given id does not exists in the db`, async () => {
    jest.spyOn(classRepository, 'findById').mockImplementationOnce(() => '');

    const response = await request(app.getHttpServer()).patch(
      '/api/v1/class/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    ); // It passes without the body, because every property is optional

    expect(response.status).toEqual(404);
    expect(response.body).toMatchSnapshot();
  });

  it(`Update class - should fail when id parameter is not UUID`, async () => {
    const response = await request(app.getHttpServer()).patch(
      '/api/v1/class/123',
    ); // It passes without the body, because every property is optional

    expect(response.status).toEqual(422);
    expect(response.body).toMatchSnapshot();
  });

  it(`Delete class - should pass when class with given id exists in the db`, async () => {
    const response = await request(app.getHttpServer()).delete(
      '/api/v1/class/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    );

    expect(response.status).toEqual(200);
    expect(response.body).toMatchSnapshot();
  });

  it(`Delete class - should fail when class with given id does not exist in the db`, async () => {
    jest.spyOn(classRepository, 'findById').mockImplementationOnce(() => '');

    const response = await request(app.getHttpServer()).delete(
      '/api/v1/class/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    );

    expect(response.status).toEqual(404);
    expect(response.body).toMatchSnapshot();
  });

  it(`Delete class - should fail when id parameter is not UUID`, async () => {
    const response = await request(app.getHttpServer()).delete(
      '/api/v1/class/123',
    );

    expect(response.status).toEqual(422);
    expect(response.body).toMatchSnapshot();
  });

  afterAll(async () => {
    await app.close();
  });
});
