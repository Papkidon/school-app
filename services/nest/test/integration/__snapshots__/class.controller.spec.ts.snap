// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ClassController] Integration tests Create class - should fail when class with given number already exists 1`] = `
{
  "message": "Class with given number already exists",
  "statusCode": 409,
}
`;

exports[`[ClassController] Integration tests Create class - should fail when request body is incorrect 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "object",
    "message": "Required",
    "path": [
      "data",
    ],
    "received": "undefined",
  },
  {
    "code": "unrecognized_keys",
    "keys": [
      "wrong",
    ],
    "message": "Unrecognized key(s) in object: 'wrong'",
    "path": [],
  },
]
`;

exports[`[ClassController] Integration tests Create class - should pass when request body is correct 1`] = `{}`;

exports[`[ClassController] Integration tests Delete class - should fail when class with given id does not exist in the db 1`] = `
{
  "message": "Class with given id does not exist in the database",
  "statusCode": 404,
}
`;

exports[`[ClassController] Integration tests Delete class - should fail when id parameter is not UUID 1`] = `
[
  {
    "code": "invalid_string",
    "message": "Invalid uuid",
    "path": [],
    "validation": "uuid",
  },
]
`;

exports[`[ClassController] Integration tests Delete class - should pass when class with given id exists in the db 1`] = `{}`;

exports[`[ClassController] Integration tests Get all classes - should fail when no class data exists in the db 1`] = `
{
  "message": "No class data exists in the database",
  "statusCode": 404,
}
`;

exports[`[ClassController] Integration tests Get all classes - should pass when at least one class exists in the db 1`] = `
[
  "dummyFoundClasses",
]
`;

exports[`[ClassController] Integration tests Get class by id - should fail when class with given id does not exists in the db 1`] = `
{
  "message": "User with given id does not exist in the database",
  "statusCode": 404,
}
`;

exports[`[ClassController] Integration tests Get class by id - should fail when id parameter is not UUID 1`] = `
[
  {
    "code": "invalid_string",
    "message": "Invalid uuid",
    "path": [],
    "validation": "uuid",
  },
]
`;

exports[`[ClassController] Integration tests Get class by id - should pass when class with given id exists in the db 1`] = `{}`;

exports[`[ClassController] Integration tests Get classes with subjects - should fail when data does not exists 1`] = `
{
  "message": "No class data exists in the database",
  "statusCode": 404,
}
`;

exports[`[ClassController] Integration tests Get classes with subjects - should pass when data exists 1`] = `
[
  {
    "id": "d379d73a-7958-44e4-ab59-0e4128f35295",
    "max_population": 20,
    "number": 211,
    "subjects": [
      {
        "id": "917d669d-ea51-4031-b7ef-25b1850bbd52",
        "name": "geography",
        "points": 1,
      },
    ],
  },
  {
    "id": "261133c7-c73c-4a7f-8199-548798dc0e97",
    "max_population": 15,
    "number": 111,
    "subjects": [
      {
        "id": "c3114bc1-1bfa-4f44-ba17-f199a6dc635c",
        "name": "math",
        "points": 5,
      },
    ],
  },
]
`;

exports[`[ClassController] Integration tests Update class - should fail when class with given id does not exists in the db 1`] = `
{
  "message": "Class with given id does not exist in the database",
  "statusCode": 404,
}
`;

exports[`[ClassController] Integration tests Update class - should fail when id parameter is not UUID 1`] = `
[
  {
    "code": "invalid_string",
    "message": "Invalid uuid",
    "path": [],
    "validation": "uuid",
  },
]
`;

exports[`[ClassController] Integration tests Update class - should pass when class with given id exists in the db 1`] = `{}`;
